io/src/real_sense_2_grabber.cpp

```diff
@@ -0,0 +1,394 @@
+        auto ptr = vertices_ptr + index;
+        auto uvptr = texture_ptr + index;
+        auto& p = cloud->points[index];
+
+        p.x = ptr->x;
+        p.y = ptr->y;
+        p.z = ptr->z;
+
+        p.intensity = getTextureIntensity(ir, uvptr->u, uvptr->v);
+
+      }
+    }
+
+
+
+    return cloud;
+  }
+
+  pcl::PointCloud<pcl::PointXYZRGB>::Ptr 
+  RealSense2Grabber::convertRGBDepthToPointXYZRGB (const rs2::points & points, rs2::video_frame & rgb)
+  {
+    pcl::PointCloud<pcl::PointXYZRGB>::Ptr cloud (new pcl::PointCloud<pcl::PointXYZRGB> ());
+
+    auto sp = points.get_profile ().as<rs2::video_stream_profile> ();
+    cloud->width = sp.width ();
+    cloud->height = sp.height ();
+    cloud->is_dense = false;
+    cloud->points.resize (points.size ());
+
+    auto vertices_ptr = points.get_vertices ();
+    auto texture_ptr = points.get_texture_coordinates ();
+
+#ifdef _OPENMP
+#pragma omp parallel for 
+#endif
+    for (int index = 0; index < cloud->points.size (); ++index)
+    {
+      auto ptr = vertices_ptr + index;
+      auto uvptr = texture_ptr + index;
+      auto& p = cloud->points[index];
+
+      p.x = ptr->x;
+      p.y = ptr->y;
+      p.z = ptr->z;
+
+      std::tie (p.r, p.g, p.b) = getTextureColor (rgb, uvptr->u, uvptr->v);
+
+      p.a = 255;
+    }
+
+    return cloud;
+  }
+
+  pcl::PointCloud<pcl::PointXYZRGBA>::Ptr 
+  RealSense2Grabber::convertRGBADepthToPointXYZRGBA (const rs2::points & points, rs2::video_frame & rgb)
+  {
+    pcl::PointCloud<pcl::PointXYZRGBA>::Ptr cloud (new pcl::PointCloud<pcl::PointXYZRGBA> ());
+
+    auto sp = points.get_profile ().as<rs2::video_stream_profile> ();
+    cloud->width = sp.width ();
+    cloud->height = sp.height ();
+    cloud->is_dense = false;
+    cloud->points.resize (points.size ());
+
+    auto vertices_ptr = points.get_vertices ();
+    auto texture_ptr = points.get_texture_coordinates ();
+
+#ifdef _OPENMP
+#pragma omp parallel for 
+#endif
+    for (int index = 0; index < cloud->points.size (); ++index)
+    {
+      auto ptr = vertices_ptr + index;
+      auto uvptr = texture_ptr + index;
+      auto& p = cloud->points[index];
+
+      p.x = ptr->x;
+      p.y = ptr->y;
+      p.z = ptr->z;
+
+      std::tie (p.r, p.g, p.b) = getTextureColor (rgb, uvptr->u, uvptr->v);
+
+      p.a = 255;
+    }
+
+    return cloud;
+  }
+
+  std::tuple<uint8_t, uint8_t, uint8_t> 
+  RealSense2Grabber::getTextureColor (rs2::video_frame &texture, float u, float v)
+  {
+    auto ptr = dynamic_cast<rs2::video_frame*>(&texture);
+
+    const int w = ptr->get_width (), h = ptr->get_height ();
+    int x = std::min (std::max (int (u*w + .5f), 0), w - 1);
+    int y = std::min (std::max (int (v*h + .5f), 0), h - 1);
+    int idx = x * ptr->get_bytes_per_pixel() + y * ptr->get_stride_in_bytes ();
+    const auto texture_data = reinterpret_cast<const uint8_t*>(ptr->get_data ());
+    return std::make_tuple (texture_data[idx], texture_data[idx + 1], texture_data[idx + 2]);
+  }
```

Comment:
The requested item was to make texture const. 

1. You're converting `rs2::video_frame&` to a `rs2::video_frame*` for no apparent reason. Every time `ptr` is used is to invoke a method (which you could do directly with the reference)
2. All methods you invoke on `texture` are getters which means they should be also declared `const`, unless someone overlooked that during the API development of RealSense.

Reply:
```cpp
std::tuple<uint8_t, uint8_t, uint8_t> 
getTextureColor (const rs2::video_frame& texture, const float u, const float v) const;

uint8_t
getTextureIntensity (const rs2::video_frame& texture, const float u, const float v) const;
```
```cpp
std::tuple<uint8_t, uint8_t, uint8_t> 
RealSense2Grabber::getTextureColor (const rs2::video_frame& texture, const float u, const float v) const
{
  const int w = texture.get_width (), h = texture.get_height ();
  const int x = std::min (std::max (static_cast<int> (u * w + .5f), 0), w - 1);
  const int y = std::min (std::max (static_cast<int> (v * h + .5f), 0), h - 1);
  const int idx = x * texture.get_bytes_per_pixel() + y * texture.get_stride_in_bytes ();
  const auto texture_data = reinterpret_cast<const uint8_t*> (texture.get_data ());
  return std::make_tuple (texture_data[idx], texture_data[idx + 1], texture_data[idx + 2]);
}

uint8_t
RealSense2Grabber::getTextureIntensity (const rs2::video_frame& texture, const float u, const float v) const
{
  const int w = texture.get_width (), h = texture.get_height ();
  const int x = std::min (std::max (static_cast<int> (u * w + .5f), 0), w - 1);
  const int y = std::min (std::max (static_cast<int> (v * h + .5f), 0), h - 1);
  const int idx = x * texture.get_bytes_per_pixel () + y * texture.get_stride_in_bytes ();
  const auto texture_data = reinterpret_cast<const uint8_t*> (texture.get_data ());
  return texture_data[idx];
}
```